src/oop/object_interactions/hard/q3/MainTest.java

```diff
@@ -0,0 +1,232 @@
+    private ObjectTest classInstance;
+    private ObjectTest testInstance;
+    private final String objectClassName = "BloodDonor";
+    private final String testClassName = "TestBlood";
+    private final String stringAttribute1 = "name";
+    private final String stringAttribute2 = "bloodType";
+    private final String getStringAttribute1 = "getName";
+    private final String getStringAttribute2 = "getBloodType";
+    private final String methodName1 = "canReceiveBloodFrom";
+    private final String testString1 = "Jill";
+    private final String testString2 = "A";
+    private final String testString3 = "Jane";
+    private final String testString4 = "B";
+    private final String testStringMessage = "They are incompatible";
+
+    @BeforeEach
+    public void setup() {
+        String objectString = "oop.object_interactions.hard.q3." + objectClassName;
+        classInstance = new ObjectTest(objectString);
+        String testString = "oop.object_interactions.hard.q3." + testClassName;
+        testInstance = new ObjectTest(testString);
+    }
+
+    @Test
+    public void bloodDonorClassHasRequiredAttributes() {
+        assertTrue(classInstance.hasField(stringAttribute1, String.class),
+                String.format("We expected an attribute named %s in your code but it was not found or was misspelled.", stringAttribute1));
+        assertTrue(classInstance.hasField(stringAttribute2, String.class),
+                String.format("We expected an attribute named %s in your code but it was not found or was misspelled.", stringAttribute2));
+        assertTrue(classInstance.hasModifier(stringAttribute1, "private"),
+                String.format("The %s field does not have the correct visibility modifier.", stringAttribute1));
+        assertTrue(classInstance.hasModifier(stringAttribute2, "private"),
+                String.format("The %s field does not have the correct visibility modifier.", stringAttribute2));
+    }
+
+    private static Stream<Arguments> constructorInputProvider() {
+        return Stream.of(
+                Arguments.of("Cindy", "A"),
+                Arguments.of("Dale", "B"),
+                Arguments.of("Iris", "AB"),
+                Arguments.of("Dave", "O")
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void bloodDonorClassHasCorrectConstructor(String name, String blood) throws Throwable {
+        Object[][] arguments = {
+                {name, String.class},
+                {blood, String.class}
+        };
+        assertTrue(classInstance.hasConstructor(new Class[]{String.class, String.class}),
+                String.format("Your %s class is missing a required constructor.", objectClassName));
+        assertTrue(classInstance.hasConstructor(new Class[]{String.class, String.class}, new String[]{"public"}),
+                String.format("Your %s class constructor has the incorrect visibility modifier.", objectClassName));
```

Comment:
Put these two in a separate test, we need to give grade just for writing constructor as well